xdsl/transforms/convert_stencil_to_csl_stencil.py

```diff
@@ -535,6 +537,37 @@ def get_prefetch_overhead(o: OpResult):
             rewriter.erase_op(prefetch.op)
 
 
+class PromoteCoefficients(RewritePattern):
+    """
+    Promotes constant coefficients to attributes. When a `csl_stencil.access` is immediately multiplied by
+    an `arith.constant` as the sole use of the accessed data, the constant is promoted to a coefficient property
+    in the `csl_stencil.apply` op.
+    """
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: csl_stencil.AccessOp, rewriter: PatternRewriter, /):
+        if (
+            not isinstance(apply := op.get_apply(), csl_stencil.ApplyOp)
+            or not op.op == apply.receive_chunk.block.args[0]
+            or len(op.result.uses) != 1
+            or not isinstance(mulf := list(op.result.uses)[0].operation, arith.Mulf)
+        ):
+            return
+
+        coeff = mulf.lhs if op.result == mulf.rhs else mulf.rhs
+
+        if (
+            not isinstance(cnst := coeff.owner, arith.Constant)
+            or not isinstance(dense := cnst.value, DenseIntOrFPElementsAttr)
```

Comment:
Can this not be a `FloatAttr`/`AnyFloatAttr`?

Reply:
It's technically checking for a tensor constant, but in a way that makes pyright happy. I'm not entirely sure what could be done with a scalar constant in this context.